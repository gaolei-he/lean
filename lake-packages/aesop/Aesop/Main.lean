/-
Copyright (c) 2021 Jannis Limperg. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Jannis Limperg
-/

import Aesop.Search.Main
import Aesop.BuiltinRules -- ensures that the builtin rules are registered
import Aesop.Frontend.Tactic
import Aesop.Profiling

open Lean
open Lean.Elab.Tactic

namespace Aesop

@[tactic Frontend.Parser.aesopTactic, tactic Frontend.Parser.aesopTactic?]
def evalAesop : Tactic := λ stx => do
  profileitM Exception "aesop" (← getOptions) do
  withMainContext do
    let (profile, totalTime) ← time do
      let (config, configParseTime) ← time $ Frontend.TacticConfig.parse stx
      let profile := { Profile.empty with configParsing := configParseTime }
      let goal ← getMainGoal
      let ((goal, ruleSet), ruleSetConstructionTime) ← time $
        config.getRuleSet goal
      let profile :=
        { profile with ruleSetConstruction := ruleSetConstructionTime }
      withConstAesopTraceNode .ruleSet (return "Rule set") do
        ruleSet.trace .ruleSet
      let (profile, searchTime) ← time do
        let (goals, profile) ←
          search goal ruleSet config.options config.simpConfig
            config.simpConfigSyntax? profile
        replaceMainGoal goals.toList
        pure profile
      pure { profile with search := searchTime }
    let profile := { profile with total := totalTime }
    profile.trace .profile

end Aesop
