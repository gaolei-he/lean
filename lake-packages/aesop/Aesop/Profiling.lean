/-
Copyright (c) 2022 Jannis Limperg. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Jannis Limperg
-/

import Aesop.Nanos
import Aesop.Tree.Data
import Aesop.Tracing

open Lean

namespace Aesop


-- All times are in nanoseconds.
structure RuleProfile where
  rule : DisplayRuleName
  elapsed : Nanos
  successful : Bool
  deriving Inhabited

namespace RuleProfile

instance : ToString RuleProfile where
  toString rp :=
    let success := if rp.successful then "successful" else "failed"
    s!"[{rp.elapsed.printAsMillis}] apply rule {rp.rule} ({success})"

end RuleProfile


-- All times are in nanoseconds.
structure Profile where
  total : Nanos
  configParsing : Nanos
  ruleSetConstruction : Nanos
  search : Nanos
  ruleSelection : Nanos
  ruleApplications : Array RuleProfile
  deriving Inhabited

namespace Profile

protected def empty : Profile where
  total := 0
  configParsing := 0
  ruleSetConstruction := 0
  search := 0
  ruleSelection := 0
  ruleApplications := #[]

instance : EmptyCollection Profile :=
  ⟨Profile.empty⟩

-- The returned map associates to each rule the total time spent on successful
-- and failed applications of that rule.
def ruleApplicationTotals (p : Profile) :
    HashMap DisplayRuleName (Nanos × Nanos) := Id.run do
  let mut m := {}
  for rp in p.ruleApplications do
    if rp.successful then
      m :=
        match m.find? rp.rule with
        | none => m.insert rp.rule (rp.elapsed, 0)
        | some (successful, failed) =>
          m.insert rp.rule (successful + rp.elapsed, failed)
    else
      m :=
        match m.find? rp.rule with
        | none => m.insert rp.rule (0, rp.elapsed)
        | some (successful, failed) =>
          m.insert rp.rule (successful, failed + rp.elapsed)
  return m

def trace (p : Profile) (opt : TraceOption) : CoreM Unit := do
  if ! (← opt.isEnabled) then
    return
  let totalRuleApplications :=
    p.ruleApplications.foldl (init := 0) λ total rp =>
      total + rp.elapsed
  aesop_trace![opt] "Total: {p.total.printAsMillis}"
  aesop_trace![opt] "Configuration parsing: {p.configParsing.printAsMillis}"
  aesop_trace![opt] "Rule set construction: {p.ruleSetConstruction.printAsMillis}"
  withConstAesopTraceNode opt (collapsed := false)
      (return m!"Search: {p.search.printAsMillis}") do
    aesop_trace![opt] "Rule selection: {p.ruleSelection.printAsMillis}"
    withConstAesopTraceNode opt (collapsed := false)
        (return m!"Rule applications: {totalRuleApplications.printAsMillis}") do
      let timings :=
        p.ruleApplicationTotals.fold
          (init := Array.mkEmpty p.ruleApplicationTotals.size)
          λ timings n (successful, failed) =>
            timings.push (n, successful, failed)
      let timings := timings.qsortOrd (ord := ⟨compareTimings⟩)
      for (n, s, f) in timings do
        aesop_trace![opt] "[{(s + f).printAsMillis} / {s.printAsMillis} / {f.printAsMillis}] {n}"
  where
    compareTimings (x y : DisplayRuleName × Nanos × Nanos) : Ordering :=
      compareLex
        (compareOn (λ (_, s, f) => s + f))
        (compareOn (λ (n, _, _) => n))
        x y
      |>.swap

end Profile


abbrev ProfileT m := StateRefT' IO.RealWorld Profile m

namespace ProfileT

protected def run [Monad m] [MonadLiftT (ST IO.RealWorld) m] (x : ProfileT m α)
    (profile : Profile) : m (α × Profile) :=
  StateRefT'.run x profile

-- Can this be expressed in terms of the various monad classes?
def liftBase [MonadLiftT m n] (x : ProfileT m α) : ProfileT n α :=
  λ r => x r

end ProfileT


class abbrev MonadProfile (m : Type → Type _) :=
  MonadOptions m
  MonadStateOf Profile m

variable [Monad m] [MonadProfile m]

@[inline, always_inline]
def isProfilingEnabled [MonadProfile m] : m Bool :=
  return (← getOptions).getBool `profiler

@[inline, always_inline]
def recordRuleSelectionProfile (elapsed : Nanos) : m Unit :=
  modify λ p => { p with ruleSelection := p.ruleSelection + elapsed }

@[inline, always_inline]
def recordRuleProfile (rp : RuleProfile) : m Unit :=
  modify λ p => { p with ruleApplications := p.ruleApplications.push rp }

@[inline, always_inline]
def profiling [MonadLiftT BaseIO m] (x : m α)
    (recordProfile : α → Nanos → m Unit) : m α := do
  if ← isProfilingEnabled then
    let (result, elapsed) ← time x
    recordProfile result elapsed
    return result
  else
    x

end Aesop
