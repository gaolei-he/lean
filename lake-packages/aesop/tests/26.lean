/-
Copyright (c) 2022 Jannis Limperg. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Asta H. From, Jannis Limperg
-/

import Aesop

set_option aesop.check.all true

attribute [-simp] List.all_cons List.all_nil List.all_eq_true

theorem all_cons (P : α → Bool) (x : α) (xs : List α) (h : (x :: xs).all P)
  : P x ∧ xs.all P := by
  aesop
