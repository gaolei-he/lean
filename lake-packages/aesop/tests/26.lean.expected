tests/26.lean:15:2: warning: aesop: failed to prove the goal after exhaustive search.
tests/26.lean:14:22: error: unsolved goals
case left
α : Type u_1
P : α → Bool
x : α
xs : List α
h : List.all (x :: xs) P = true
⊢ P x = true

case right
α : Type u_1
P : α → Bool
x : α
xs : List α
h : List.all (x :: xs) P = true
⊢ List.all xs P = true
